llvm/lib/Transforms/Instrumentation/AddressSanitizer.cpp

```diff
@@ -1230,7 +1234,92 @@ static bool isUnsupportedAMDGPUAddrspace(Value *Addr) {
+  if (PtrTy->getPointerAddressSpace() != 0) {
+    // Skip SPIR-V built-in varibles
+    StringRef Name;
+    if (auto *CE = dyn_cast<ConstantExpr>(Addr)) {
+      const auto OpCode = CE->getOpcode();
+      if (OpCode == Instruction::GetElementPtr ||
+          OpCode == Instruction::AddrSpaceCast) {
+        Name = CE->getOperand(0)->getName();
+      }
+    } else {
+      Name = Addr->getName();
+    }
+    if (Name.starts_with("__spirv_BuiltIn")) {
+      return true;
+    }
+  } else {
+    // FIXME: Check kernel argments instead of value's name
+    auto Name = Addr->getName();
+    if (Name == "_arg_NumWorkItems" || Name == "_arg_KernelFunc") {
+      return true;
+    }
+  }
+  return false;
+}
+
+static GlobalVariable *GetOrCreateGlobalString(Module &M, StringRef Name,
+                                               StringRef String,
+                                               unsigned AddressSpace) {
+  static std::unordered_map<std::string, GlobalVariable *> GlobalStringMap;
+  GlobalVariable *StringGV = nullptr;
+  if (GlobalStringMap.find(String.str()) != GlobalStringMap.end()) {
+    StringGV = GlobalStringMap[String.str()];
+  } else {
+    auto *Ty =
+        ArrayType::get(Type::getInt8Ty(M.getContext()), String.size() + 1);
+    StringGV = new GlobalVariable(
+        M, Ty, true, GlobalValue::InternalLinkage,
+        ConstantDataArray::getString(M.getContext(), String), Name, nullptr,
+        GlobalValue::NotThreadLocal, AddressSpace);
+    GlobalStringMap[String.str()] = StringGV;
+  }
+  return StringGV;
+}
+
+static void AppendDebugInfoToArgs(Instruction *InsertBefore, Value *Addr,
+                                  SmallVector<Value *> &Args) {
+  auto *M = InsertBefore->getModule();
+  auto &C = InsertBefore->getContext();
+  auto &DB = InsertBefore->getDebugLoc();
+  constexpr unsigned ConstantAS = 2;
+  auto *I8PtrTy = Type::getInt8Ty(C)->getPointerTo(ConstantAS);
+
+  // Address Space
+  Type *PtrTy = cast<PointerType>(Addr->getType()->getScalarType());
+  Args.push_back(
+      ConstantInt::get(Type::getInt32Ty(C), PtrTy->getPointerAddressSpace()));
+
+  // File & Line
+  if (DB) {
+    auto FileName = DB->getFilename();
+    auto *FileNameGV =
+        GetOrCreateGlobalString(*M, "__asan_file", FileName, ConstantAS);
+    Args.push_back(ConstantExpr::getPointerCast(FileNameGV, I8PtrTy));
+    Args.push_back(ConstantInt::get(Type::getInt32Ty(C), DB.getLine()));
+  } else {
+    Args.push_back(ConstantPointerNull::get(I8PtrTy));
+    Args.push_back(ConstantInt::get(Type::getInt32Ty(C), 0));
+  }
+
+  // Function
+  auto FuncName = InsertBefore->getFunction()->getName();
+  auto *FuncNameGV = GetOrCreateGlobalString(*M, "__asan_func",
+                                             demangle(FuncName), ConstantAS);
+  Args.push_back(ConstantExpr::getPointerCast(FuncNameGV, I8PtrTy));
+}
+
 Value *AddressSanitizer::memToShadow(Value *Shadow, IRBuilder<> &IRB) {
+  if (TargetTriple.isSPIR()) {
```

Comment:
this check might not be accurate since there could be users of this pass in a case of spir target other than sycl.
Should other spir targets also use this code path?
Is there a way to determine that the code is for intel gpu only?

Reply:
> Should other spir targets also use this code path?

Does OpenMP also use "spir" target?

> Is there a way to determine that the code is for intel gpu only?

This code is not intel gpu only, ocl cpu also needs it.

Reply:
Do we expect that non-intel gpu/cpu also use it? I think we probably need a more fine-grained control so that this code is only used for intel targets.

Reply:
> Do we expect that non-intel gpu/cpu also use it? I think we probably need a more fine-grained control so that this code is only used for intel targets.

Good points, I'll investigate this.

Reply:
Since intel/llvm is specific to the intel compiler, so I think it's okay here? Unless this modification is conflict with other products.